README.md
# Brainless Token Management :)

[![npm version](https://badge.fury.io/js/brainless-token-manager.svg)](https://badge.fury.io/js/brainless-token-manager) [![npm](https://img.shields.io/npm/dt/brainless-token-manager.svg?logo=npm)](https://www.npmjs.com/package/brainless-token-manager) [![npm](https://img.shields.io/bundlephobia/minzip/brainless-token-manager)](https://www.npmjs.com/package/brainless-token-manager)
[![All Contributors](https://img.shields.io/badge/all_contributors-1-orange.svg)](#contributors-)

[Live Demo](https://reactjs-handle-refresh-token.vercel.app/)

## Installation

[![NPM](https://nodei.co/npm/brainless-token-manager.png?compact=true)](https://nodei.co/npm/brainless-token-manager/)

## To install the latest stable version:

```
npm install --save brainless-token-manager@latest

or

yarn add brainless-token-manager
```

## Introduction

- This package help you do refresh token brainlessly

## Flow

- Checking refresh token -> expired -> onInvalidRefreshToken -> clear token on your storage -> logout
- Valid token -> return token -> run as normal
- Token in valid -> refresh token -> onRefreshToken success -> save token and refresh token to storage -> perform request

## Super easy to use

### API

```typescript
// Works fine with JWT
// if you use other tokens JWT. you need to initialize isValidToken and isValidRefreshToken
interface TokenManagerContructor {
  getAccessToken: () => Promise<string>;
  getRefreshToken: () => Promise<string>; // if you don't have refresh token use the same as getAccessToken
  executeRefreshToken?: () => Promise<{ token: string; refresh_token: string }>;
  onRefreshTokenSuccess?: ({ token, refresh_token }: { token: string; refresh_token: string }) => void;
  onInvalidRefreshToken: () => void; // will trigger when refresh token expired
  isValidToken?: (token: string) => Promise<boolean>;
  isValidRefreshToken?: (refresh_token: string) => Promise<boolean>;
  refreshTimeout?: number;
}

const tokenManagerInstance: TokenManager = new TokenManager(options: TokenManagerContructor);
```

### Flow

![Demo](./img/image.png)

### Example with umi-request

```javascript
import { extend } from 'umi-request';
import TokenManager, { injectBearer, parseJwt } from 'brainless-token-manager';

// Can implement by umi-request, axios, fetch....
export const requestNew = extend({
  prefix: 'APP_ENDPOINT_URL_HERE',
  headers: {
    'Content-Type': 'application/json',
  },
  errorHandler: (error) => {
    throw error?.data || error?.response;
  },
});

const tokenManager = new TokenManager({
  getAccessToken: async () => {
    const token = localStorage.getItem('accessToken');
    return `${token}`;
  },
  getRefreshToken: async () => {
    const refreshToken = localStorage.getItem('refreshToken');

    return `${refreshToken}`;
  },
  onInvalidRefreshToken: () => {
    // Logout, redirect to login
    localStorage.removeItem('accessToken');
    localStorage.removeItem('refreshToken');
  },
  executeRefreshToken: async () => {
    const refreshToken = localStorage.getItem('refreshToken');

    if (!refreshToken) {
      return {
        token: '',
        refresh_token: '',
      };
    }

    const r = await requestNew.post('/auth/refresh-token', {
      data: {
        refreshToken: refreshToken,
      },
    });

    return {
      token: r?.accessToken,
      refresh_token: r?.refreshToken,
    };
  },
  onRefreshTokenSuccess: ({ token, refresh_token }) => {
    if (token && refresh_token) {
      localStorage.setItem('accessToken', token);
      localStorage.setItem('refreshToken', refresh_token);
    }
  },
});

export const privateRequest = async (request: any, suffixUrl: string, configs?: any) => {
  const token: string = await tokenManager.getToken();

  return request(suffixUrl, injectBearer(token, configs));
};

// Use
privateRequest(axios.get, '/example', { params: { 'work': 'ReactNative' } } });
```

## Compare

- TokenManager

![Token manager](./assets/token-manager.gif)

- Axios Interceptor

![Token manager](./assets/axios.gif)


CHANGELOG.md
# brainless-token-manager

## 1.3.3

### Patch Changes

- 91081a5: add error logger

## 1.3.1

### Patch Changes

- c9737d3: Update readme

## 1.3.0

### Minor Changes

- c8202d9: onRefreshTokenSuccess and executeRefreshToken as optional, use in no refresh token scenario

## 1.2.14

### Patch Changes

- f918c6c: Support RN

## 1.2.13

### Patch Changes

- 6c3f647: Update atob

## 1.2.12

### Patch Changes

- c2c6cc7: update documentation

## 1.2.11

### Patch Changes

- bb32c04: Update readme

## 1.2.10

### Patch Changes

- 35dadd1: Update type

## 1.2.8

### Patch Changes

- 8bc5f5e: Update ci

## 1.2.6

### Patch Changes

- b4b11da: Update document

## 1.2.4

### Patch Changes

- 654ab2b: update readme and package


src/index.ts
import EventEmitter from './EventEmitter';

export interface TokenManagerContructor {
  getAccessToken: () => Promise<string>;
  getRefreshToken: () => Promise<string>;
  isValidToken?: (token: string) => Promise<boolean>;
  isValidRefreshToken?: (refresh_token: string) => Promise<boolean>;
  executeRefreshToken?: () => Promise<{ token: string; refresh_token: string }>;
  onRefreshTokenSuccess?: ({ token, refresh_token }: { token: string; refresh_token: string }) => void;
  onInvalidRefreshToken: () => void;
  refreshTimeout?: number;
}

export interface TokenManagerInstance {
  getToken: () => Promise<string>;
}

export type TInjectBearerConfigs = {
  headers?: { Authorization?: string } & {
    [K in string]: any;
  };
};

export type TParseJwt = { exp: number } & {
  [K in string]: any;
};

enum ERROR {
  ERROR_PARSER_TOKEN = 'ERROR_PARSER_TOKEN',
  NOT_FOUND_EXPIRE_TIME = 'NOT_FOUND_EXPIRE_TIME',
}

export const parseJwt = (token: string): Partial<TParseJwt> => {
  try {
    return JSON.parse(Base64.atob(token.split('.')[1]));
  } catch (e) {
    console.error(ERROR.ERROR_PARSER_TOKEN);
    return {};
  }
};

export const injectBearer = <TToken extends string, TConfig extends TInjectBearerConfigs>(
  token: TToken,
  configs?: TConfig,
): TConfig => {
  if (!configs) {
    return {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    } as TConfig;
  }

  if (configs?.headers?.Authorization) {
    return {
      ...configs,
      headers: {
        ...configs.headers,
      },
    };
  }

  if (configs?.headers) {
    return {
      ...configs,
      headers: {
        ...configs.headers,
        Authorization: `Bearer ${token}`,
      },
    };
  }

  return {
    ...configs,
    headers: {
      Authorization: `Bearer ${token}`,
    },
  };
};

export default class TokenManager {
  private event: EventEmitter;
  public getAccessToken;
  public getRefreshToken;
  private onInvalidRefreshToken;
  private isRefreshing: boolean = false;
  private refreshTimeout: number = 30000;
  private isValidRefreshToken;
  private onRefreshTokenSuccess;
  private isValidToken;
  private executeRefreshToken;

  constructor({
    getRefreshToken,
    getAccessToken,
    isValidToken,
    refreshTimeout = 30000,
    executeRefreshToken,
    onInvalidRefreshToken,
    onRefreshTokenSuccess,
    isValidRefreshToken,
  }: TokenManagerContructor) {
    const event = new EventEmitter();
    this.refreshTimeout = refreshTimeout;
    this.getAccessToken = getAccessToken;
    this.getRefreshToken = getRefreshToken;
    this.onInvalidRefreshToken = onInvalidRefreshToken;
    this.onRefreshTokenSuccess = onRefreshTokenSuccess;
    this.executeRefreshToken = executeRefreshToken;

    if (isValidToken) {
      this.isValidToken = isValidToken;
    } else {
      this.isValidToken = this.isTokenValid;
    }

    if (isValidRefreshToken) {
      this.isValidRefreshToken = isValidRefreshToken;
    } else {
      this.isValidRefreshToken = this.isTokenValid;
    }

    event.on('refreshTokenExpired', () => {
      this.onInvalidRefreshToken && this.onInvalidRefreshToken();
    });

    event.on('refresh', () => {
      (async () => {
        try {
          const refreshToken = await this.getRefreshToken();
          const isRefreshTokenValid = await this.isValidRefreshToken(refreshToken);
          if (!isRefreshTokenValid) {
            event.emit('refreshTokenExpired');
          } else {
            const token = await getAccessToken();
            const isValid: boolean = await this.isValidToken(token);
            if (isValid) {
              event.emit('refreshDone', token);
            } else {
              event.emit('refreshing');
            }
          }
        } catch (e) {}
      })();
    });

    event.on('refreshing', async () => {
      if (this.isRefreshing) {
        return;
      }

      if (!this.executeRefreshToken) {
        return;
      }
      // fetch
      this.isRefreshing = true;

      const evtFire = false;
      const { token, refresh_token } = await this.executeRefreshToken();

      if (token && refresh_token) {
        this.onRefreshTokenSuccess && this.onRefreshTokenSuccess({ token, refresh_token });
      }
      this.event.emit('refreshDone', token);
      this.isRefreshing = false;

      if (this.refreshTimeout) {
        setTimeout(() => {
          if (!evtFire) {
            this.event.emit('refreshDone', null);
            this.isRefreshing = false;
          }
        }, this.refreshTimeout);
      }
    });

    this.event = event;
  }

  getToken(): Promise<string | unknown> {
    return new Promise((resolve) => {
      let isCalled = false;

      const refreshDoneHandler = (token: string) => {
        resolve(token);
        isCalled = true;
      };

      this.event.once('refreshDone', refreshDoneHandler);

      if (!isCalled) {
        this.event.emit('refresh');
      }
    });
  }

  async isTokenValid(token: string) {
    try {
      const decoded = parseJwt(token);

      if (!decoded?.exp) {
        console.error(ERROR.NOT_FOUND_EXPIRE_TIME);

        return true;
      }

      const currentTime = Date.now() / 1000;

      if (decoded.exp - 5 > currentTime) {
        return true;
      }

      return false;
    } catch (error) {
      return false;
    }
  }
}

const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
const Base64 = {
  btoa: (input: string = '') => {
    let str = input;
    let output = '';

    for (
      let block = 0, charCode, i = 0, map = chars;
      str.charAt(i | 0) || ((map = '='), i % 1);
      output += map.charAt(63 & (block >> (8 - (i % 1) * 8)))
    ) {
      charCode = str.charCodeAt((i += 3 / 4));

      if (charCode > 0xff) {
        throw new Error("'btoa' failed: The string to be encoded contains characters outside of the Latin1 range.");
      }

      block = (block << 8) | charCode;
    }

    return output;
  },

  atob: (input: string = '') => {
    let str = input.replace(/=+$/, '');
    let output = '';

    if (str.length % 4 == 1) {
      throw new Error("'atob' failed: The string to be decoded is not correctly encoded.");
    }
    for (
      let bc = 0, bs = 0, buffer, i = 0;
      (buffer = str.charAt(i++));
      ~buffer && ((bs = bc % 4 ? bs * 64 + buffer : buffer), bc++ % 4)
        ? (output += String.fromCharCode(255 & (bs >> ((-2 * bc) & 6))))
        : 0
    ) {
      buffer = chars.indexOf(buffer);
    }

    return output;
  },
};


lib/index.d.ts
export interface TokenManagerContructor {
  getAccessToken: () => Promise<string>;
  getRefreshToken: () => Promise<string>;
  isValidToken?: (token: string) => Promise<boolean>;
  isValidRefreshToken?: (refresh_token: string) => Promise<boolean>;
  executeRefreshToken: () => Promise<{
    token: string;
    refresh_token: string;
  }>;
  onRefreshTokenSuccess: ({ token, refresh_token }: { token: string; refresh_token: string }) => void;
  onInvalidRefreshToken: () => void;
  refreshTimeout?: number;
}

export type TInjectBearerConfigs = {
  headers?: {
    Authorization?: string;
  } & {
    [K in string]: any;
  };
};
export type TParseJwt = {
  exp: number;
} & {
  [K in string]: any;
};
export declare const parseJwt: (token: string) => Partial<TParseJwt>;
export declare const injectBearer: <TToken extends string, TConfig extends TInjectBearerConfigs>(
  token: TToken,
  configs?: TConfig | undefined,
) => TConfig;
export default class TokenManager {
  private event;
  getAccessToken: () => Promise<string>;
  getRefreshToken: () => Promise<string>;
  private onInvalidRefreshToken;
  private isRefreshing;
  private refreshTimeout;
  private isValidRefreshToken;
  private onRefreshTokenSuccess;
  private isValidToken;
  constructor({
    getRefreshToken,
    getAccessToken,
    isValidToken,
    refreshTimeout,
    executeRefreshToken,
    onInvalidRefreshToken,
    onRefreshTokenSuccess,
    isValidRefreshToken,
  }: TokenManagerContructor);
  getToken(): Promise<string | unknown>;
  isTokenValid(token: string): Promise<boolean>;
}


lib/EventEmitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EventEmitter = /** @class */ (function () {
    function EventEmitter() {
        this.events = {};
    }
    EventEmitter.prototype._getEventListByName = function (eventName) {
        if (typeof this.events[eventName] === 'undefined') {
            this.events[eventName] = new Set();
        }
        return this.events[eventName];
    };
    EventEmitter.prototype.on = function (eventName, fn) {
        var _a;
        (_a = this._getEventListByName(eventName)) === null || _a === void 0 ? void 0 : _a.add(fn);
    };
    EventEmitter.prototype.once = function (eventName, fn) {
        var _this = this;
        var onceFn = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            _this.removeListener(eventName, onceFn);
            fn.apply(_this, args);
        };
        this.on(eventName, onceFn);
    };
    EventEmitter.prototype.emit = function (eventName) {
        var _this = this;
        var _a;
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        (_a = this._getEventListByName(eventName)) === null || _a === void 0 ? void 0 : _a.forEach(function (fn) {
            fn === null || fn === void 0 ? void 0 : fn.apply(_this, args);
        });
    };
    EventEmitter.prototype.removeListener = function (eventName, fn) {
        var _a;
        (_a = this._getEventListByName(eventName)) === null || _a === void 0 ? void 0 : _a.delete(fn);
    };
    return EventEmitter;
}());
exports.default = EventEmitter;


lib/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.injectBearer = exports.parseJwt = void 0;
var EventEmitter_1 = require("./EventEmitter");
var parseJwt = function (token) {
    try {
        return JSON.parse(atob(token.split('.')[1]));
    }
    catch (e) {
        return {};
    }
};
exports.parseJwt = parseJwt;
var injectBearer = function (token, configs) {
    var _a;
    if (!configs) {
        return {
            headers: {
                Authorization: "Bearer ".concat(token),
            },
        };
    }
    if ((_a = configs === null || configs === void 0 ? void 0 : configs.headers) === null || _a === void 0 ? void 0 : _a.Authorization) {
        return __assign(__assign({}, configs), { headers: __assign({}, configs.headers) });
    }
    if (configs === null || configs === void 0 ? void 0 : configs.headers) {
        return __assign(__assign({}, configs), { headers: __assign(__assign({}, configs.headers), { Authorization: "Bearer ".concat(token) }) });
    }
    return __assign(__assign({}, configs), { headers: {
            Authorization: "Bearer ".concat(token),
        } });
};
exports.injectBearer = injectBearer;
var TokenManager = /** @class */ (function () {
    function TokenManager(_a) {
        var getRefreshToken = _a.getRefreshToken, getAccessToken = _a.getAccessToken, isValidToken = _a.isValidToken, _b = _a.refreshTimeout, refreshTimeout = _b === void 0 ? 30000 : _b, executeRefreshToken = _a.executeRefreshToken, onInvalidRefreshToken = _a.onInvalidRefreshToken, onRefreshTokenSuccess = _a.onRefreshTokenSuccess, isValidRefreshToken = _a.isValidRefreshToken;
        var _this = this;
        this.isRefreshing = false;
        this.refreshTimeout = 30000;
        var event = new EventEmitter_1.default();
        this.refreshTimeout = refreshTimeout;
        this.getAccessToken = getAccessToken;
        this.getRefreshToken = getRefreshToken;
        this.onInvalidRefreshToken = onInvalidRefreshToken;
        this.onRefreshTokenSuccess = onRefreshTokenSuccess;
        if (isValidToken) {
            this.isValidToken = isValidToken;
        }
        else {
            this.isValidToken = this.isTokenValid;
        }
        if (isValidRefreshToken) {
            this.isValidRefreshToken = isValidRefreshToken;
        }
        else {
            this.isValidRefreshToken = this.isTokenValid;
        }
        event.on('refreshTokenExpired', function () {
            _this.onInvalidRefreshToken && _this.onInvalidRefreshToken();
        });
        event.on('refresh', function () {
            (function () { return __awaiter(_this, void 0, void 0, function () {
                var refreshToken, isRefreshTokenValid, token, isValid, e_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, 7, , 8]);
                            return [4 /*yield*/, this.getRefreshToken()];
                        case 1:
                            refreshToken = _a.sent();
                            return [4 /*yield*/, this.isValidRefreshToken(refreshToken)];
                        case 2:
                            isRefreshTokenValid = _a.sent();
                            if (!!isRefreshTokenValid) return [3 /*break*/, 3];
                            event.emit('refreshTokenExpired');
                            return [3 /*break*/, 6];
                        case 3: return [4 /*yield*/, getAccessToken()];
                        case 4:
                            token = _a.sent();
                            return [4 /*yield*/, this.isValidToken(token)];
                        case 5:
                            isValid = _a.sent();
                            if (isValid) {
                                event.emit('refreshDone', token);
                            }
                            else {
                                event.emit('refreshing');
                            }
                            _a.label = 6;
                        case 6: return [3 /*break*/, 8];
                        case 7:
                            e_1 = _a.sent();
                            return [3 /*break*/, 8];
                        case 8: return [2 /*return*/];
                    }
                });
            }); })();
        });
        event.on('refreshing', function () { return __awaiter(_this, void 0, void 0, function () {
            var evtFire, _a, token, refresh_token;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (this.isRefreshing) {
                            return [2 /*return*/];
                        }
                        // fetch
                        this.isRefreshing = true;
                        evtFire = false;
                        return [4 /*yield*/, executeRefreshToken()];
                    case 1:
                        _a = _b.sent(), token = _a.token, refresh_token = _a.refresh_token;
                        if (token && refresh_token) {
                            this.onRefreshTokenSuccess && this.onRefreshTokenSuccess({ token: token, refresh_token: refresh_token });
                        }
                        this.event.emit('refreshDone', token);
                        this.isRefreshing = false;
                        if (this.refreshTimeout) {
                            setTimeout(function () {
                                if (!evtFire) {
                                    _this.event.emit('refreshDone', null);
                                    _this.isRefreshing = false;
                                }
                            }, this.refreshTimeout);
                        }
                        return [2 /*return*/];
                }
            });
        }); });
        this.event = event;
    }
    TokenManager.prototype.getToken = function () {
        var _this = this;
        return new Promise(function (resolve) {
            var isCalled = false;
            var refreshDoneHandler = function (token) {
                resolve(token);
                isCalled = true;
            };
            _this.event.once('refreshDone', refreshDoneHandler);
            if (!isCalled) {
                _this.event.emit('refresh');
            }
        });
    };
    TokenManager.prototype.isTokenValid = function (token) {
        return __awaiter(this, void 0, void 0, function () {
            var decoded, exp, currentTime;
            return __generator(this, function (_a) {
                try {
                    decoded = (0, exports.parseJwt)(token);
                    exp = decoded.exp;
                    currentTime = Date.now() / 1000;
                    if (exp && exp - 5 > currentTime) {
                        return [2 /*return*/, true];
                    }
                    return [2 /*return*/, false];
                }
                catch (error) {
                    return [2 /*return*/, false];
                }
                return [2 /*return*/];
            });
        });
    };
    return TokenManager;
}());
exports.default = TokenManager;
